### ISSUE TITLE
CRC32 performance similar to JS

### ISSUE THREAD

**Issue by 101arrowz** [Daninet/hash-wasm#13]
First off, I wanted to thank you for releasing `hash-wasm` for free; it's very well made and is one of the few libraries I've seen that both has a nice API and high performance. I really appreciate the Base64 builds by default; apparently, not enough WASM library authors understand how painful it is to get WASM working in bundlers :)

I'm considering using the CRC32 implementation from `hash-wasm` as an optional extension in [fflate](https://github.com/101arrowz/fflate) for GZIP and ZIP compression. However, I'd like to clarify a few things before I add official support.

In the benchmark, `hash-wasm`'s CRC32, which uses Slice-by-8, is compared to byte-iterative implementations, which are obviously going to be much slower. Slice-by-8 and Slice-by-16 CRC32 in JS are actually quite similar in performance to `hash-wasm`, if the benchmark is to be believed. I added this to `crc32.worker.js`:

```js
// CRC32 table
const crct = (() => {
  const t = new Int32Array(4096);
  for (let i = 0; i < 256; ++i) {
    let c = i,
      k = 9;
    while (--k) c = (c & 1 && -306674912) ^ (c >>> 1);
    t[i] = c;
  }
  for (let i = 0; i < 256; ++i) {
    let lv = t[i];
    for (let j = 256; j < 4096; j += 256)
      lv = t[i | j] = (lv >>> 8) ^ t[lv & 255];
  }
  return t;
})();

const t1 = crct.slice(0, 256),
  t2 = crct.slice(256, 512),
  t3 = crct.slice(512, 768),
  t4 = crct.slice(768, 1024);
const t5 = crct.slice(1024, 1280),
  t6 = crct.slice(1280, 1536),
  t7 = crct.slice(1536, 1792),
  t8 = crct.slice(1792, 2048);
const t9 = crct.slice(2048, 2304),
  t10 = crct.slice(2304, 2560),
  t11 = crct.slice(2560, 2816),
  t12 = crct.slice(2816, 3072);
const t13 = crct.slice(3072, 3328),
  t14 = crct.slice(3328, 3584),
  t15 = crct.slice(3584, 3840),
  t16 = crct.slice(3840, 4096);

suite.addSync(`hyper`, (d) => {
  let cr = -1,
    i = 0;
  if (d.length > 256) {
    const max = d.length - 16;
    for (; i < max; ) {
      cr =
        t16[d[i++] ^ (cr & 255)] ^
        t15[d[i++] ^ ((cr >> 8) & 255)] ^
        t14[d[i++] ^ ((cr >> 16) & 255)] ^
        t13[d[i++] ^ (cr >>> 24)] ^
        t12[d[i++]] ^
        t11[d[i++]] ^
        t10[d[i++]] ^
        t9[d[i++]] ^
        t8[d[i++]] ^
        t7[d[i++]] ^
        t6[d[i++]] ^
        t5[d[i++]] ^
        t4[d[i++]] ^
        t3[d[i++]] ^
        t2[d[i++]] ^
        t1[d[i++]];
    }
  }
  for (; i < d.length; ++i) cr = t1[(cr & 255) ^ d[i]] ^ (cr >>> 8);
  cr = ~cr;
  if (cr < 0) return (0x100000000 + cr).toString(16);
  return cr.toString(16);
});
```

And here are my results:
![Screenshot](https://user-images.githubusercontent.com/29579245/112354538-04de8700-8c8a-11eb-8c46-380f2cd3dc59.png)

Performance is within 10% for the 1MB variant, which for `fflate` isn't worth the extra bundle size that the WASM adds. Bundle size could be improved by computing the tables at runtime, at which point it is probably worth at least supporting `hash-wasm`, but the CRC32 implementation in `hash-wasm` would need to be either smaller or significantly (>30%) faster to justify using it by default. My devices are all quite powerful, so perhaps it's a rare case. Maybe you could run the benchmark to see if this JS version is actually fast or if it's just for me?

If you find that the JS version tends to be slower on most devices, or if you find a way to improve CRC32 performance, I'll look into implementing `hash-wasm` as the default for `fflate` with large files in async mode. Thanks again for creating this awesome library!

**Comment by Daninet** [comment]
Thank you for the kind words. 😊

On my computer (Win 10, Chrome x64 v89, I7-7700K, DevTools closed) I have these numbers:

![image](https://user-images.githubusercontent.com/884882/112365074-e4d9b280-8cdf-11eb-84e7-14779ca671c8.png)

With DevTools open, I get half as good performance, something similar to what you see. This is a known problem of the V8 engine. Some optimization levels are disabled if the inspection mode is turned on. 

In general, I could not archive very good speedups at the CRC32 algorithm, especially when the input data is small. If you check in Firefox, the performance profile of the CRC32 algorithm is very different:
![image](https://user-images.githubusercontent.com/884882/112366103-14d58580-8ce1-11eb-8afd-cf49ae61d87e.png)
You can see that for small inputs, `hash-wasm` is beaten by all JS implementations.

**Comment by Daninet** [comment]
`Bundle size could be improved by computing the tables at runtime` -> that's a good idea, I will try measure it this weekend.

**Comment by 101arrowz** [comment]
I'm mainly concerned about mid-size inputs (always bigger than 32 bytes, almost always smaller than 1MB). I'm looking for the optimal chunk size at which `wasm-flate` is faster. For inputs under around 32kB, I was planning on using JS by default since WASM will be at best slightly faster and at worst much slower depending on the engine. Above that, I'd think maybe WASM could have a performance advantage? But then there's also the memory concerns to consider; since V8 will be started anyway, the memory used by WASM (i.e. when receiving the data to hash) is worse. I'll add a wiki entry in any case.

**Comment by Daninet** [comment]
I released a new version (4.6.0) in which I compute the lookup table on the first run. I did some measurements and it has insignificant impact on performance while the CRC32 bundle size decreased to 3kB.

In the long run I believe the WASM will be faster in all scenarios and engines. They put a great effort into optimizing the compilers, because a lot of major players like Autodesk ported their native apps to WASM and they need performance.
Recent news from the V8 team: https://v8.dev/blog/v8-release-90#faster-js-to-wasm-calls

Memory shouldn't be a concern in my opinion: It only allocates 128kB RAM for the WASM VM and the JS <-> WASM data transfer is through a 16kB data buffer.

**Comment by 101arrowz** [comment]
Thanks! I've already seen the V8 release, but given how few calls there are to WASM from `hash-wasm`, I doubt there'll be a noticeable performance bump. It's very useful for rapid information interchange, however.

I didn't realize that you were chunking the data transfer to WASM, so at least memory usage will be fine. (Smart decision I probably wouldn't have considered 😁)

After evaluating the performance difference between the 1-byte CRC algorithm and the slice-by-N algorithms, the extra bundle size of ~0.8kB with the Slice-by-16 JS version and 2.2kB for the Slice-by-8 WASM version isn't worth it for most users (CRC32 takes around 1% of the time of GZIP compression; making that 0.5% doesn't change much). So I'll add `hash-wasm` to a wiki entry under a "performance optimizations" section for those who are compressing a lot of files. Thanks again for making this library :)

P.S. While I agree that the performance of WASM is quite good, I'm not so sure if it applies in all circumstances. For example, `wasm-flate` is quite slow, much slower than `fflate` by about 2x despite being a WASM build of native Zlib. That's why I thought memory copying when moving to WASM might be a concern, since it seems to be the only way Zlib could be slower.
